File: ChessApp/src/CSharp9-DSL/ChessApp/ChessApp_Internals/ConsoleUI_Internals/CommandParser.cs
using System;
using System.Linq;
using ChessApp.ChessApp_Internals.ConsoleUI_Internals.Commands;
using HLCD.Infrastructure;
using Pidgin;
using static Pidgin.Parser;
using static Pidgin.Parser<char>;

namespace ChessApp.ChessApp_Internals.ConsoleUI_Internals
{
    [Component("CA-CUI-CP")]
    class CommandParser
    {
        public CommandParser()
        {
            Parser = CreateParser();
        }

        public string GetHelp() =>
@"
Game commands:
    save <game name>
    load <game name>
    del  <game name>
    list            - lists saved games
    new <color>     - starts new game, playing with black or white
    exit            - exit
    
Turns:
    <from cell> - <to cell>   - move or eat
        e.g.: E2 - E4
    
    <from cell> - <to figure> - pawn transformation
        e.g.: D7 - Queen
    
    castle - <to cell>        - castle
        e.g.: castle - A7
    
    draw                      - offer/accept draw
    reject                    - reject draw
";
        private Parser<char, Command> Parser { get; }

        private Parser<char, Command> CreateParser()
        {
            var ws = Whitespaces;
            var color = TryEnum<Color>();
            var figure = TryEnum<Figure>();

            var name =
                from first in Token(char.IsLetter)
                from rest in Token(char.IsLetterOrDigit).ManyString()
                select first + rest;

            var cell =
                from h in Token(c => c >= Board.Left && c <= Board.Right)
                from v in Token(c => c >= '1' && c <= '8')
                select Cell.At(h, int.Parse($"{v}"));

            var exit = Try(String("exit")).ThenReturn(Command.Exit).Cast<Command>();
            var list = Try(String("list")).ThenReturn(Command.List).Cast<Command>();
            var newGame = Try(String("new")).Then(ws).Then(color).Select(Command.NewGame).Cast<Command>();
            var saveGame = Try(String("save")).Then(ws).Then(name).Select(Command.Save).Cast<Command>();
            var loadGame = Try(String("load")).Then(ws).Then(name).Select(Command.Load).Cast<Command>();
            var delGame = Try(String("del")).Then(ws).Then(name).Select(Command.Delete).Cast<Command>();

            var move = Try(
                from fromCell in cell
                from toCell in ws.Then(Char('-')).Then(ws).Then(cell)
                select new Move(fromCell, toCell))
                .Cast<Turn>();

            var turn = OneOf(move).Select(t => new TurnCommand(t)).Cast<Command>();

            return OneOf(exit, list, newGame, saveGame, loadGame, delGame, turn);
        }

        private Parser<char, TEnum> TryEnum<TEnum>() where TEnum : Enum
            => OneOf(
                Enum.GetValues(typeof(TEnum))
                    .Cast<TEnum>()
                    .Select(v => Try(String(Enum.GetName(typeof(TEnum), v)!)).ThenReturn(v))
                    .ToArray());

        public Command Parse(string cmdStr)
        {
            cmdStr = cmdStr.Trim();
            var result = Parser.Parse(cmdStr);
            return result.Success
                ? result.Value
                : throw new UserError($"Bad command: {cmdStr}, {result.Error!}");
        }
    }
}

File: ChessApp/src/RustMacros/src/chess_app/console_ui/command_parser.rs
#[cfg(test)]
mod tests;

pub mod interface {
    use std::fmt::Display;
    use super::super::data::command::Command;
    
    #[derive(PartialEq, Eq, Clone, Debug)]
    pub struct Error(pub String); 
    
    impl std::error::Error for Error{
    }
    
    impl Display for Error {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            write!(f, "{}", self.0)
        }
    }
    
    hlcd::define! {
        interface CommandParser {
            fn parse(&self, s: &str) -> Result<Command, Error>;
            fn get_help(&self) -> String;
        }
    }
}

pub mod component {
    use nom::{
        IResult, 
        branch::alt, 
        bytes::complete::tag_no_case,    
        combinator::{map, recognize}, 
        sequence::separated_pair, 
        character::complete::{alpha1, space1}, 
        multi::many1};
    
    use crate::{chess_app::data::{Color, Turn}};
    use nom_extensions::parseable::Parseable;
    use super::{interface::*, super::data::command::Command};
    
    hlcd::define! {
        component CommandParser {
            provides { CommandParser }
    
            impl {
                fn ident(input: &str) -> IResult<&str, &str> {
                    recognize(many1(alpha1))(input)
                }
            
                fn parse<'a>(&self, input: &'a str) -> IResult<&'a str, Command> {
                    use Command::*;
                    let ident = Self::ident;
                    let color = Color::nom_parse;
            
                    let exit = map(tag_no_case("exit"), |_| Exit);
                    let help = map(tag_no_case("help"), |_| Help);
                    let list = map(tag_no_case("list"), |_| ListGames);
            
                    let load = map(separated_pair(tag_no_case("load"), space1, ident), 
                        |(_, name)| LoadGame(name.to_string()));
                    
                    let save = map(separated_pair(tag_no_case("save"), space1, ident), 
                        |(_, name)| SaveGame(name.to_string()));
                    
                    let del = map(separated_pair(tag_no_case("del"), space1, ident), 
                        |(_, name)| DeleteGame(name.to_string()));
                    
                    let new = map(separated_pair(tag_no_case("new"), space1, color), 
                        |(_, color)| NewGame(color));
            
                    let turn = map(Turn::nom_parse,
                        |turn| MakeTurn(turn)); 
            
                    alt((exit, help, list, load, save, del, new, turn))
                    (input)
                }    
            }
    
            impl CommandParser {
                fn parse(&self, s: &str) -> Result<Command, Error> {
                    match self.parse(s) {
                        Ok((_, cmd)) => Ok(cmd),
                        Err(e) => Err(Error(format!("Invalid command. {:?}", e)))
                    }
                }
            
                fn get_help(&self) -> String {
                    r#"
                        Game commands:
                            save <game name>
                            load <game name>
                            del  <game name>
                            list            - lists saved games
                            new <color>     - starts new game, playing with black or white
                            exit            - exit
                            
                        Turns:
                            <from cell> - <to cell>   - move or eat
                                e.g.: E2 - E4
                            
                            <from cell> - <to figure> - pawn transformation
                                e.g.: D7 - Queen
                            
                            castle - <to cell>        - castle
                                e.g.: castle - A7
                            
                            draw                      - offer/accept draw
                            reject                    - reject draw
                                
                    "#.to_string()
                }        
            }
        }
    }
}
